File: packages/stores/services/dynamodb.test.ts
import { describe, expect, test } from 'bun:test';
import { LEASE_MS, storeContractSuite, T0, TTL_MS } from '@anyonce/core/testing';
import {
  DescribeTimeToLiveCommand,
  DynamoDBClient,
  GetItemCommand,
} from '@aws-sdk/client-dynamodb';
import { DYNAMODB_MAX_RESULT_BYTES, DynamoDbStore, ensureTable, itemKey } from '../src/dynamodb';
import { describeService } from './services';

const TABLE = `anyonce_test_${Date.now()}`;

function client(): DynamoDBClient {
  return new DynamoDBClient({
    region: 'us-east-1',
    endpoint: 'http://127.0.0.1:18000',
    credentials: { accessKeyId: 'local', secretAccessKey: 'local' },
  });
}

await describeService('dynamodb store', 18000, () => {
  const shared = client();
  const ready = ensureTable(shared, TABLE);

  storeContractSuite(
    'dynamodb',
    async () => {
      await ready;
      const store = new DynamoDbStore({ client: shared, tableName: TABLE });
      return {
        store,
        physicallyRemove: (op) => store.physicallyRemove(op),
        maxResultBytes: DYNAMODB_MAX_RESULT_BYTES,
      };
    },
    { describe, test, expect },
    { maxResultBytes: DYNAMODB_MAX_RESULT_BYTES, nativePurge: true },
  );

  describe('dynamodb specifics', () => {
    test('REQ-ST-DDB-1: a refused begin classifies the outcome from ReturnValuesOnConditionCheckFailure with no second call', async () => {
      await ready;
      let commands: string[] = [];
      const counting = client();
      counting.middlewareStack.add(
        (next, context) => async (args) => {
          commands.push(context.commandName ?? 'unknown');
          return next(args);
        },
        { step: 'initialize', name: 'count', priority: 'low' },
      );
      const store = new DynamoDbStore({ client: counting, tableName: TABLE });
      const op = { scope: `rvocf-${Date.now()}`, key: 'k', fingerprint: 'a' };
      await store.begin(op, { leaseMs: LEASE_MS, ttlMs: TTL_MS, now: T0 });
      await store.complete(op, 1, { kind: 'http', status: 200, body: new Uint8Array([1]) }, T0 + 1);
      commands = [];
      const mismatch = await store.begin(
        { ...op, fingerprint: 'b' },
        { leaseMs: LEASE_MS, ttlMs: TTL_MS, now: T0 + 2 },
      );
      expect(mismatch.outcome).toBe('mismatch');
      expect(commands).toEqual(['UpdateItemCommand']);
      commands = [];
      const completed = await store.begin(op, { leaseMs: LEASE_MS, ttlMs: TTL_MS, now: T0 + 3 });
      expect(completed.outcome).toBe('completed');
      if (completed.outcome === 'completed')
        expect(completed.record.result?.body).toEqual(new Uint8Array([1]));
      expect(commands).toEqual(['UpdateItemCommand']);
      counting.destroy();
    });

    test('REQ-ST-DDB-1: the ttl attribute is enabled on the table and set relative to the wall clock plus the grace', async () => {
      await ready;
      const ttlSpec = await shared.send(new DescribeTimeToLiveCommand({ TableName: TABLE }));
      expect(ttlSpec.TimeToLiveDescription?.TimeToLiveStatus).toBe('ENABLED');
      expect(ttlSpec.TimeToLiveDescription?.AttributeName).toBe('ttl');
      const store = new DynamoDbStore({
        client: shared,
        tableName: TABLE,
        nativeTtlGraceMs: 60_000,
      });
      const op = { scope: `ttl-${Date.now()}`, key: 'k', fingerprint: 'a' };
      const before = Date.now();
      await store.begin(op, { leaseMs: LEASE_MS, ttlMs: 5_000, now: T0 });
      const item = await shared.send(
        new GetItemCommand({
          TableName: TABLE,
          Key: { pk: { S: itemKey(op) } },
          ConsistentRead: true,
        }),
      );
      const ttl = Number(item.Item?.ttl?.N);
      expect(ttl).toBeGreaterThan(Math.floor((before + 5_000 + 60_000) / 1000) - 2);
      expect(ttl).toBeLessThan(Math.floor((before + 5_000 + 60_000) / 1000) + 5);
      expect(Number(item.Item?.expires_at?.N)).toBe(T0 + 5_000);
    });

    test('REQ-ST-DDB-1: the item lives under the single partition key pk, the scope and key joined (Q22)', async () => {
      await ready;
      const store = new DynamoDbStore({ client: shared, tableName: TABLE });
      const op = { scope: `q22-${Date.now()}`, key: 'k/with/slashes', fingerprint: 'a' };
      await store.begin(op, { leaseMs: LEASE_MS, ttlMs: TTL_MS, now: T0 });
      const composite = `${op.scope}${String.fromCharCode(31)}${op.key}`;
      expect(itemKey(op)).toBe(composite);
      const found = await shared.send(
        new GetItemCommand({
          TableName: TABLE,
          Key: { pk: { S: composite } },
          ConsistentRead: true,
        }),
      );
      expect(found.Item?.pk?.S).toBe(composite);
      expect(found.Item?.sk).toBeUndefined();
      expect(Number(found.Item?.fence?.N)).toBe(1);
      // The row still decodes back into its scope and key.
      const record = await store.get(op, T0 + 1);
      expect(record?.scope).toBe(op.scope);
      expect(record?.key).toBe(op.key);
    });

    test('REQ-ST-DDB-1: purge is a no-op that returns 0 and ensureTable is idempotent', async () => {
      await ready;
      const store = new DynamoDbStore({ client: shared, tableName: TABLE });
      expect(await store.purge(Date.now())).toBe(0);
      await ensureTable(shared, TABLE);
    });
  });
});

File: go/store/dynamodb/dynamodb_test.go
package dynamodb_test

import (
	"context"
	"fmt"
	"net/http"
	"testing"
	"time"

	"github.com/aws/aws-sdk-go-v2/aws"
	"github.com/aws/aws-sdk-go-v2/credentials"
	awsdynamodb "github.com/aws/aws-sdk-go-v2/service/dynamodb"
	"github.com/aws/aws-sdk-go-v2/service/dynamodb/types"
	"github.com/sns45/anyonce/go/anyonce"
	"github.com/sns45/anyonce/go/conformance"
	"github.com/sns45/anyonce/go/conformance/fixture"
	"github.com/sns45/anyonce/go/httpmw"
	"github.com/sns45/anyonce/go/store/dynamodb"
	"github.com/sns45/anyonce/go/store/internal/servicetest"
	"github.com/sns45/anyonce/go/storetest"
)

func client() *awsdynamodb.Client {
	return awsdynamodb.New(awsdynamodb.Options{
		Region:       "us-east-1",
		BaseEndpoint: aws.String("http://127.0.0.1:18000"),
		Credentials:  credentials.NewStaticCredentialsProvider("local", "local", ""),
	})
}

func TestDynamoDBStore(t *testing.T) {
	servicetest.Require(t, "dynamodb", "127.0.0.1:18000")
	ctx := context.Background()
	c := client()
	table := fmt.Sprintf("anyonce_go_%d", time.Now().UnixNano())
	if err := dynamodb.EnsureTable(ctx, c, table); err != nil {
		t.Fatal(err)
	}
	storetest.Run(t, "dynamodb", func(*testing.T) storetest.Harness {
		s := dynamodb.New(c, dynamodb.Options{Table: table})
		return storetest.Harness{Store: s, PhysicallyRemove: s.PhysicallyRemove, MaxResultBytes: dynamodb.MaxResultBytes, NativePurge: true}
	})

	t.Run("REQ-ST-DDB-1: EnsureTable is idempotent, taking the ResourceInUse and already-enabled TTL branches", func(t *testing.T) {
		if err := dynamodb.EnsureTable(ctx, c, table); err != nil {
			t.Fatalf("second EnsureTable on the same table: %v", err)
		}
	})

	t.Run("REQ-ST-DDB-1: a refused begin classifies from the returned old item and the ttl attribute is enabled", func(t *testing.T) {
		s := dynamodb.New(c, dynamodb.Options{Table: table})
		op := anyonce.Operation{Scope: fmt.Sprintf("rvocf-%d", time.Now().UnixNano()), Key: "k", Fingerprint: "a"}
		opts := anyonce.BeginOptions{Lease: storetest.Lease, TTL: storetest.TTL, Now: storetest.T0}
		if _, err := s.Begin(ctx, op, opts); err != nil {
			t.Fatal(err)
		}
		if _, err := s.Complete(ctx, op, 1, anyonce.StoredResult{Kind: anyonce.KindHTTP, Status: 200, Body: []byte{1}}, storetest.T0.Add(time.Millisecond)); err != nil {
			t.Fatal(err)
		}
		out, err := s.Begin(ctx, anyonce.Operation{Scope: op.Scope, Key: op.Key, Fingerprint: "b"}, opts)
		if err != nil || out.Kind != anyonce.BeginMismatch || out.Record == nil || out.Record.Result == nil || string(out.Record.Result.Body) != "\x01" {
			t.Fatalf("%+v %v", out, err)
		}
		ttl, err := c.DescribeTimeToLive(ctx, &awsdynamodb.DescribeTimeToLiveInput{TableName: aws.String(table)})
		if err != nil || ttl.TimeToLiveDescription.TimeToLiveStatus != types.TimeToLiveStatusEnabled || *ttl.TimeToLiveDescription.AttributeName != "ttl" {
			t.Fatalf("%+v %v", ttl, err)
		}
		if n, err := s.Purge(ctx, time.Now()); err != nil || n != 0 {
			t.Fatalf("purge %d %v", n, err)
		}
	})

	t.Run("REQ-ST-DDB-1: the native ttl attribute is the wall clock plus the ttl plus the grace, in seconds", func(t *testing.T) {
		s := dynamodb.New(c, dynamodb.Options{Table: table, NativeTTLGrace: 90 * time.Second})
		op := anyonce.Operation{Scope: fmt.Sprintf("ttlattr-%d", time.Now().UnixNano()), Key: "k", Fingerprint: "a"}
		before := time.Now()
		if _, err := s.Begin(ctx, op, anyonce.BeginOptions{Lease: storetest.Lease, TTL: time.Hour, Now: storetest.T0}); err != nil {
			t.Fatal(err)
		}
		item, err := c.GetItem(ctx, &awsdynamodb.GetItemInput{
			TableName:      aws.String(table),
			Key:            map[string]types.AttributeValue{"pk": &types.AttributeValueMemberS{Value: dynamodb.ItemKey(op.Scope, op.Key)}},
			ConsistentRead: aws.Bool(true),
		})
		if err != nil {
			t.Fatal(err)
		}
		ttlAttr, ok := item.Item["ttl"].(*types.AttributeValueMemberN)
		if !ok {
			t.Fatalf("no numeric ttl attribute: %+v", item.Item)
		}
		var seconds int64
		if _, err := fmt.Sscan(ttlAttr.Value, &seconds); err != nil {
			t.Fatal(err)
		}
		low, high := before.Add(time.Hour+90*time.Second).Unix(), time.Now().Add(time.Hour+90*time.Second).Unix()
		if seconds < low || seconds > high {
			t.Fatalf("ttl %d outside [%d, %d]", seconds, low, high)
		}
		// The logical expiry comes from opts.Now, never from the wall clock.
		expires, ok := item.Item["expires_at"].(*types.AttributeValueMemberN)
		if !ok || expires.Value != fmt.Sprint(storetest.T0.Add(time.Hour).UnixMilli()) {
			t.Fatalf("expires_at %+v", item.Item["expires_at"])
		}
	})

	t.Run("REQ-ST-DDB-1: the item lives under the single partition key pk, the scope and key joined (Q22)", func(t *testing.T) {
		s := dynamodb.New(c, dynamodb.Options{Table: table})
		op := anyonce.Operation{Scope: fmt.Sprintf("q22-%d", time.Now().UnixNano()), Key: "k/with/slashes", Fingerprint: "a"}
		if _, err := s.Begin(ctx, op, anyonce.BeginOptions{Lease: storetest.Lease, TTL: storetest.TTL, Now: storetest.T0}); err != nil {
			t.Fatal(err)
		}
		composite := op.Scope + string(rune(31)) + op.Key
		if dynamodb.ItemKey(op.Scope, op.Key) != composite {
			t.Fatalf("ItemKey is %q", dynamodb.ItemKey(op.Scope, op.Key))
		}
		item, err := c.GetItem(ctx, &awsdynamodb.GetItemInput{
			TableName:      aws.String(table),
			Key:            map[string]types.AttributeValue{"pk": &types.AttributeValueMemberS{Value: composite}},
			ConsistentRead: aws.Bool(true),
		})
		if err != nil {
			t.Fatal(err)
		}
		pk, ok := item.Item["pk"].(*types.AttributeValueMemberS)
		if !ok || pk.Value != composite {
			t.Fatalf("pk %+v", item.Item["pk"])
		}
		if _, hasSort := item.Item["sk"]; hasSort {
			t.Fatal("the item still carries a sort key")
		}
		rec, err := s.Get(ctx, op.Scope, op.Key, storetest.T0.Add(time.Millisecond))
		if err != nil || rec == nil || rec.Scope != op.Scope || rec.Key != op.Key {
			t.Fatalf("%+v %v", rec, err)
		}
	})

	t.Run("REQ-ST-DDB-1: every core and profile vector passes through httpmw with the DynamoDB store", func(t *testing.T) {
		f := fixture.New()
		// No MaxResultBytes in the policy on purpose: the store implements anyonce.ResultCapper, so httpmw caps
		// the policy at it and the run proves the automatic cap rather than an explicit override.
		mw := httpmw.New(dynamodb.New(c, dynamodb.Options{Table: table}), httpmw.Options{Required: true, Policy: anyonce.Policy{TTL: 2 * time.Second}})
		mux := http.NewServeMux()
		mux.Handle("POST /reset", f.Handler())
		mux.Handle("/", mw.Handler(f.Handler()))
		summary := conformance.Run(t, mux, conformance.Options{Capabilities: []string{"short-ttl"}})
		if summary.Passed != len(summary.Results) || len(summary.Results) != 20 {
			t.Fatalf("passed %d of %d", summary.Passed, len(summary.Results))
		}
	})
}
